### ISSUE TITLE
Keras LSTM uni & bidirectional models converted to Tensorflowjs not producing correct inference

### ISSUE THREAD

**Issue by jaygit123** [tensorflow/tfjs#1220]
[`tensorflow.js`]

#### TensorFlow.js version
tensorflow                1.12.0          
tensorflow-base           1.12.0          
tensorflow-gpu            1.12.0                 
tensorflow-hub            0.2.0                    
tensorflowjs              0.8.0                     

#### Browser version
Firefox 65.0 (64-it) on Windows 10
Microsoft Edge 42.17134.1.0 on Windows 10

#### Describe the problem or feature request
I have created & trained a Keras based LSTM bidirectional model in Python to classify video. This model works awesome and classifies the videos with 90+ accuracy. **But when I converted this model to tensoflorjs model using the tensorflorjs_converter tool and used the same on browser, the model always throws the same output (top 3 results) for any video input** - BasketballDunk; prob. 0.860, BalanceBeam; prob. 0.088, BodyWeightSquats; prob. 0.024

I have checked all the inputs, their shape, etc. that are given to the LSTM bidirectional model and can't find any issues. But still the inference from LSTM bidirectional model is always the same irrespective of the video input. Please help me fixing this issue. All the required details are below.

(entire model is based on the examples given in this github repository by Xianshun Chen (chen0040) ->[https://github.com/chen0040/keras-video-classifier])

Details of the model:
 - uses MobileNet model to extract features
 - uses LSTM bidirectional model to take-in extracted features and classify the video as one of 20 classes

Dataset used:
 - UCF101 - Action Recognition Data Set (http://crcv.ucf.edu/data/UCF101.php)

Tensorflowjs converted model:
 - tensorflowjs converted model, sample videos and html file to test are all in this drive location as zip file: [https://drive.google.com/open?id=1k_4xOPlTdbUJCBPFyT9zmdB3W5lYfuw0]
 - to test the model, just unzip and build using 'yarn' and run using 'yearn watch'
 - index.html has the instructions to test

**NOTE:** I have tried LSTM model (unidirectional) and same issue is with that converted model as well. Only difference is that it produces 'Billards' as the top prediction with probability over 0.95.

#### Code to reproduce the bug / link to feature request
Code & test artifacts are in a zip file at this Drive location - [https://drive.google.com/open?id=1k_4xOPlTdbUJCBPFyT9zmdB3W5lYfuw0]

#### Link to query raised in google groups: 
https://mail.google.com/mail/u/0/#inbox/FMfcgxwBVgkWRLKPRZnRPKmpLmtdtJNx

**Comment by jaygit123** [comment]
Just to add...in the same environment, I had tried converting my finetuned MobileNet Keras model to tensorflowjs. This works perfectly in the browser. 
Also, I have ensured that every individual video frame sent to LSTM model as a sequence is correct. (used MobileNet model to recognize each frame and it does correctly and hence concluding that frames sent to LSTM are perfect).
 
So, I am sure the environment, tensorflow versions and inputs given to LSTM model are perfect without any issues. Somehow LSTM based keras model converted to tensorflowjs are not working (returning the same predictions all the time). 

I am aware that LSTM models are not supported completely in Tensorflow Lite as well (certain ops are not supported yet and so tflite converter throws error when trying to convert a keras model with LSTM)...so, wondering if tensorflowjs also has similar issues specific to LSTM.

**Comment by caisq** [comment]
Hi @jaygit123 Thanks for filing the issue. I looked into the issue a little. When I tried to run the `yarn watch` commands in the zip file that you kindly provided, I ran into an error that says it can't find "...cat.jpg".

So instead of relying on your example, I did the following:
1. I loaded the model.json with `tf.loadModel()`.
2. Synthesize some artificial inputs of the required shape ([1, 7, 1000], assuming one input example) by using `tf.ones()` and `mul()`, e.g., `tf.ones([1, 7, 1000]).mul(0.5)`, varying the multiplication constant. Or use `tf.linspace()` to create a non-constant input.
3. Feed the synthesized tensors to `model.predict()` and get the output tensor value.
4. Convert the model from the TensorFlow.js format into Keras H5 format using tensorflowjs_converter
5. Load the H5 model in Python's Keras (version 2.2.2) and do the equivalent of steps 2 and 3 above, so we can compare the output tensor values between TensorFlow.js and Python Keras.

I did steps 1, 2 and 3 in both Node.js (tfjs-node) and the browser (tested in Linux + Chrome latest) and the results are all consistent, (up to small numerical precision differences). 

For example
- Input: `tf.ones([1, 7, 1000]).mul(0.5)`:
  - Output from tfjs-node: `[[0.0010137, 0.0275199, 1e-7, 0.0045571, 0, 0.0000144, 0, 2e-7, 0, 0, 0.0934894, 0, 0.0003392, 0.022769, 5e-7,0.0000631, 0, 0.0000441, 0.0000436, 0.8501457],]`
  - Output from tfjs in the browser: `[[0.0010137, 0.0275199, 1e-7, 0.0045571, 0, 0.0000144, 0, 2e-7, 0, 0, 0.0934894, 0, 0.0003392, 0.022769, 5e-7, 0.0000631, 0, 0.0000441, 0.0000436, 0.8501457],]`
  - Output from Python Keras: `[[1.0136805e-03 2.7519871e-02 1.4615649e-07 4.5570536e-03 1.1302131e-08
  1.4366335e-05 3.0925369e-09 1.5769741e-07 1.7109763e-11 2.7308724e-09
  9.3489289e-02 9.5589858e-10 3.3919929e-04 2.2769064e-02 4.5737744e-07
  6.3108637e-05 8.5282448e-11 4.4132914e-05 4.3619501e-05 8.5014582e-01]]`
- Input: `tf.ones([1, 7, 1000]).mul(0.8)`:
  - Output from tfjs-node: `[[0.0002567, 0.0058351, 2e-7, 0.0006882, 1e-7, 0.0000904, 0, 5e-7, 0, 0, 0.9658499, 0, 0.000026, 0.0020949, 5e-7, 0.0000243, 0, 0.0000039, 0.0002725, 0.024857],]`
  - Output from tfjs in the browser: `[[0.0002567, 0.0058351, 2e-7, 0.0006882, 1e-7, 0.0000904, 0, 5e-7, 0, 0, 0.9658499, 0, 0.000026, 0.0020949, 5e-7, 0.0000243, 0, 0.0000039, 0.0002725, 0.024857],]`
  - Ouptut from Python Keras: `[[2.5674232e-04 5.8350638e-03 1.6618655e-07 6.8822311e-04 6.5749276e-08
  9.0432390e-05 1.7303192e-09 4.7633387e-07 6.0754950e-11 1.4914293e-09
  9.6584988e-01 2.9829100e-10 2.5967023e-05 2.0948723e-03 4.7803093e-07
  2.4306377e-05 4.7695115e-11 3.9397628e-06 2.7245813e-04 2.4857037e-02]]`
- Input: `tf.linspace(0, 1, 7000).reshape([1, 7, 1000])`
  - Output from tfjs-node: `[[0.002744, 0.0638575, 1e-7, 0.0020577, 0, 0.0000562, 0, 1e-7, 0, 0, 0.3674918, 0, 0.0002358, 0.0068686, 3e-7, 0.0000236, 0, 0.0000213, 0.0000923, 0.5565504],]`
  - Output from tfjs in the browser: `[[0.002744, 0.0638576, 1e-7, 0.0020577, 0, 0.0000562, 0, 1e-7, 0, 0, 0.367492, 0, 0.0002358, 0.0068686, 3e-7, 0.0000236, 0, 0.0000213, 0.0000923, 0.5565503],]`
  - Output from Python Keras: `[[2.7441687e-03 6.3858047e-02 1.2151558e-07 2.0576301e-03 8.9725027e-09
  5.6177527e-05 1.5975968e-09 1.2962559e-07 1.1335909e-11 1.2362755e-09
  3.6750904e-01 1.6217674e-10 2.3585861e-04 6.8688178e-03 3.3624949e-07
  2.3574032e-05 1.9598614e-11 2.1350013e-05 9.2276088e-05 5.5653244e-01]]`

So in summary, based on the artificial inputs I tested, the model in your `lstm_bidirectional_model` folder seems to give the same outputs as Python Keras, in both Node.js and the browser. 

Am I missing anything?

**Comment by jaygit123** [comment]
Hello @caisq,
Thanks for investigating the issue. 
My apologies for missing few files in the zip file. I have updated the zip file now with all required files. So, please download the updated zip, unzip, run 'yarn' and then 'yarn watch'. Post that, follow the instructions on index.html. (zip file to download is [here]( https://drive.google.com/open?id=1JnR_tpsqdV-oEeKOVbKn2DSzWH3CFFz_ ))
Once you follow instructions in index.html to load lstm model and video, you should see seven frames from the uploaded video being displayed, their corresponding MobileNet predictions should be displayed above them, and finally LSTM model's video classification (top3 classifications) will be displayed just below the uploaded video.
You can check the Browser 'Console' as well for debug messages I have displayed, to confirm input shape, etc. 

The zip file also contains the Keras model from which I converted to tfjs model. You can find the Keras model in folder 'lstm_bidirectional_model/mobilenet-bidirectional-lstm-model.h5'. If you would like to check this, you could provide a video input (split the video into 7 frames, extract features for each using MobileNet model and then pass them to this lstm model). BTW, I have tested this Keras model and it works perfect with 90+% accuracy on the UCF-101 dataset. So, I don't think you have to check this model. If you need the output of python keras model to cross reference with output of converted tfjs model, I can send the output of python keras model for a specific set of input videos. Let me know.

The tfjs converted model should classify these videos correctly, as its Keras model did correctly: v_BodyWeightSquats_g24_c01.mp4, v_Billiards_g08_c06.mp4, v_Biking_g03_c01.mp4, v_BabyCrawling_g06_c05.mp4, v_ApplyLipstick_g01_c04.mp4, v_ApplyEyeMakeup_g24_c05.mp4

NOTE: To use other videos that came along in zip file, open index.html and replace video src from './videos/v_BodyWeightSquats_g24_c01.mp4' to appropriate file name. Whatever video is loaded in index.html MUST be used for uploading while testing. 

Please feel free to ping me for any other details/inputs you need to investigate this.

BTW, I am not sure if the way you have tested my LSTM bidirectional model is best way to reproduce the issue described. But still, you can see in your output, the probs generated for 20 classes by tfjs in browser and python keras are different for the same input. Anyways, the zip file now contains all the required files. Please check to reproduce the issue. 

NOTE: When I saw the same predictions for any video input, I thought the frame extraction from video was not proper or the features extracted from mobilenet model were incorrect. Hence I displayed the frames from video and mobilenet's prediction values. From these, I concluded that video's frame input and extracted features are correct, but somehow converted tfjs model is producing same output all the time.

Regards,
Jay

**Comment by jaygit123** [comment]
@caisq, thought this input will be helpful.

I used the '--input_format' as 'keras' while converting python keras model to tfjs model, as my lstm model uses 'keras' library outside tf and not 'tf.keras' the one inside tf.

**Comment by jaygit123** [comment]
Hi Shanqing,Hope you saw my response and additional inputs. Any updates please?Regards,JayOn 14-Feb-2019 9:51 AM, Shanqing Cai <notifications@github.com> wrote:Hi @jaygit123 Thanks for filing the issue. I looked into the issue a little. When I tried to run the yarn watch commands in the zip file that you kindly provided, I ran into an error that says it can't find "...cat.jpg".
So instead of relying on your example, I did the following:
I loaded the model.json with tf.loadModel().Synthesize some artificial inputs of the required shape ([1, 7, 1000], assuming one input example) by using tf.ones() and mul(), e.g., tf.ones([1, 7, 1000]).mul(0.5), varying the multiplication constant. Or use tf.linspace() to create a non-constant input.Feed the synthesized tensors to model.predict() and get the output tensor value.Convert the model from the TensorFlow.js format into Keras H5 format using tensorflowjs_converterLoad the H5 model in Python's Keras (version 2.2.2) and do the equivalent of steps 2 and 3 above, so we can compare the output tensor values between TensorFlow.js and Python Keras.
I did steps 1, 2 and 3 in both Node.js (tfjs-node) and the browser (tested in Linux + Chrome latest) and the results are all consistent, (up to small numerical precision differences).
For example
Input: tf.ones([1, 7, 1000]).mul(0.5):
Output from tfjs-node: [[0.0010137, 0.0275199, 1e-7, 0.0045571, 0, 0.0000144, 0, 2e-7, 0, 0, 0.0934894, 0, 0.0003392, 0.022769, 5e-7,0.0000631, 0, 0.0000441, 0.0000436, 0.8501457],]Output from tfjs in the browser: [[0.0010137, 0.0275199, 1e-7, 0.0045571, 0, 0.0000144, 0, 2e-7, 0, 0, 0.0934894, 0, 0.0003392, 0.022769, 5e-7, 0.0000631, 0, 0.0000441, 0.0000436, 0.8501457],]Output from Python Keras: [[1.0136805e-03 2.7519871e-02 1.4615649e-07 4.5570536e-03 1.1302131e-08 1.4366335e-05 3.0925369e-09 1.5769741e-07 1.7109763e-11 2.7308724e-09 9.3489289e-02 9.5589858e-10 3.3919929e-04 2.2769064e-02 4.5737744e-07 6.3108637e-05 8.5282448e-11 4.4132914e-05 4.3619501e-05 8.5014582e-01]]
Input: tf.ones([1, 7, 1000]).mul(0.8):
Output from tfjs-node: [[0.0002567, 0.0058351, 2e-7, 0.0006882, 1e-7, 0.0000904, 0, 5e-7, 0, 0, 0.9658499, 0, 0.000026, 0.0020949, 5e-7, 0.0000243, 0, 0.0000039, 0.0002725, 0.024857],]Output from tfjs in the browser: [[0.0002567, 0.0058351, 2e-7, 0.0006882, 1e-7, 0.0000904, 0, 5e-7, 0, 0, 0.9658499, 0, 0.000026, 0.0020949, 5e-7, 0.0000243, 0, 0.0000039, 0.0002725, 0.024857],]Ouptut from Python Keras: [[2.5674232e-04 5.8350638e-03 1.6618655e-07 6.8822311e-04 6.5749276e-08 9.0432390e-05 1.7303192e-09 4.7633387e-07 6.0754950e-11 1.4914293e-09 9.6584988e-01 2.9829100e-10 2.5967023e-05 2.0948723e-03 4.7803093e-07 2.4306377e-05 4.7695115e-11 3.9397628e-06 2.7245813e-04 2.4857037e-02]]
Input: tf.linspace(0, 1, 7000).reshape([1, 7, 1000])
Output from tfjs-node: [[0.002744, 0.0638575, 1e-7, 0.0020577, 0, 0.0000562, 0, 1e-7, 0, 0, 0.3674918, 0, 0.0002358, 0.0068686, 3e-7, 0.0000236, 0, 0.0000213, 0.0000923, 0.5565504],]Output from tfjs in the browser: [[0.002744, 0.0638576, 1e-7, 0.0020577, 0, 0.0000562, 0, 1e-7, 0, 0, 0.367492, 0, 0.0002358, 0.0068686, 3e-7, 0.0000236, 0, 0.0000213, 0.0000923, 0.5565503],]Output from Python Keras: [[2.7441687e-03 6.3858047e-02 1.2151558e-07 2.0576301e-03 8.9725027e-09 5.6177527e-05 1.5975968e-09 1.2962559e-07 1.1335909e-11 1.2362755e-09 3.6750904e-01 1.6217674e-10 2.3585861e-04 6.8688178e-03 3.3624949e-07 2.3574032e-05 1.9598614e-11 2.1350013e-05 9.2276088e-05 5.5653244e-01]]

So in summary, based on the artificial inputs I tested, the model in your lstm_bidirectional_model folder seems to give the same outputs as Python Keras, in both Node.js and the browser.
Am I missing anything?

**Comment by jaygit123** [comment]
Hi @caisq, 
Any updates please? 

Regards, Jay

**Comment by jaygit123** [comment]
Hi @caisq,
Found out the reasons for tfjs converted model not producing the correct inference...at last :)

Reasons: 
1. Input to LSTM model had NaN in them! Though I was passing the extracted features from MobileNet model to LSTM, features.**dataSync() was not used**. Because of this, when I added the extracted features into a tf.buffer they were added as NaN. (when I printed values in log just before adding to tf.buffer, they printed values correctly!...this is strange). So, when I used dataSync() on the extracted features, they got added into tf.buffer correctly.
2. **Use of tf.buffer() to store the extracted features** (from MobileNet) and converting them to tensors before passing to LSTM model. Instead I used tf.stack() to store the extracted features and then passed the stacked tensor to LSTM model. (I understand that tf.stack() does the equivalent of np.array())

Thanks anyways for your time. 

BTW, I noticed some minor differences in predictions between python model and tfjs model on few occasions. Is this supposed to happen? or am I missing something? 

Regards,
Jay
